ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -2389,6 +2390,11 @@ private RelNode applyPostJoinOrderingTransform(RelNode basePlan, RelMetadataProv
 
       final HepProgramBuilder program = new HepProgramBuilder();
 
+      if (conf.getBoolVar(ConfVars.HIVE_CARDINALITY_PRESERVING_JOIN_OPTIMIZATION)) {
+        generatePartialProgram(program, false, HepMatchOrder.TOP_DOWN,
+            new HiveCardinalityPreservingJoinRule());
```

Comment:
nit. Create a static final instance and use it (as other rules do similarly).

Reply:
`HiveCardinalityPreservingJoinRule` is not stateless: it extends `HiveFieldTrimmerRule` which tracks whether it was triggered or not to avoid unnecessary multiple traversals.